runners/samza/src/test/java/org/apache/beam/runners/samza/runtime/PortableBundleManagerTest.java

```diff
@@ -0,0 +1,176 @@
+    portableBundleManager.tryStartBundle();
+    portableBundleManager.tryStartBundle();
+    portableBundleManager.tryStartBundle();
+    portableBundleManager.tryStartBundle();
+    portableBundleManager.tryFinishBundle(emitter);
+
+    verify(bundleProgressListener, times(1)).onBundleStarted();
+    verify(bundleProgressListener, times(1)).onBundleFinished(any());
+  }
+
+  @Test
+  public void testWhenBundleTimeReachedThenFinishBundle() throws Exception {
+    portableBundleManager =
+        new PortableBundleManager<>(bundleProgressListener, 4, 1, bundleTimerScheduler, TIMER_ID);
+    portableBundleManager.tryStartBundle();
+    Thread.sleep(2);
+    portableBundleManager.tryFinishBundle(emitter);
+
+    verify(bundleProgressListener, times(1)).onBundleStarted();
+    verify(bundleProgressListener, times(1)).onBundleFinished(any());
+  }
+
+  @Test
+  public void testWhenSignalFailureThenResetBundle() throws Exception {
+    portableBundleManager =
+        new PortableBundleManager<>(bundleProgressListener, 4, 1, bundleTimerScheduler, TIMER_ID);
+    portableBundleManager.tryStartBundle();
+    portableBundleManager.signalFailure(new Exception());
+    portableBundleManager.tryStartBundle();
+
+    verify(bundleProgressListener, times(2)).onBundleStarted();
+  }
+
+  @Test
+  public void testProcessWatermarkWhenBundleNotStarted() {
+    Instant watermark = new Instant();
+    portableBundleManager =
+        new PortableBundleManager<>(bundleProgressListener, 4, 1, bundleTimerScheduler, TIMER_ID);
+    portableBundleManager.processWatermark(watermark, emitter);
+    verify(bundleProgressListener, times(1)).onWatermark(eq(watermark), eq(emitter));
+  }
+
+  @Test
+  public void testQueueWatermarkWhenBundleNotStarted() {
+    Instant watermark = new Instant();
+    portableBundleManager =
+        new PortableBundleManager<>(bundleProgressListener, 1, 1, bundleTimerScheduler, TIMER_ID);
+
+    portableBundleManager.tryStartBundle();
+    portableBundleManager.processWatermark(watermark, emitter);
+    verify(bundleProgressListener, times(0)).onWatermark(eq(watermark), eq(emitter));
+
+    portableBundleManager.tryFinishBundle(emitter);
+    verify(bundleProgressListener, times(1)).onWatermark(eq(watermark), eq(emitter));
+  }
+
+  @Test
+  public void testProcessTimerTriesFinishBundle() {
+    portableBundleManager =
+        new PortableBundleManager<>(bundleProgressListener, 1, 1, bundleTimerScheduler, TIMER_ID);
+
+    portableBundleManager.tryStartBundle();
+    KeyedTimerData<Void> keyedTimerData = mock(KeyedTimerData.class);
+    TimerInternals.TimerData timerData = mock(TimerInternals.TimerData.class);
+    when(keyedTimerData.getTimerData()).thenReturn(timerData);
+    when(timerData.getTimerId()).thenReturn(TIMER_ID);
+
+    portableBundleManager.processTimer(keyedTimerData, emitter);
+    verify(bundleProgressListener, times(1)).onBundleFinished(any());
+  }
```

Comment:
can we validate if we are scheduling the next timer as well?

Reply:
Added